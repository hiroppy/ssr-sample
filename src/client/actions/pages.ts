import { State } from '../reducers/pages';

export type Actions =
  | ReturnType<typeof setEnv>
  | ReturnType<typeof setBaseUrl>
  | ReturnType<typeof resetPageStatus>
  | ReturnType<typeof loadAppProcess>
  | ReturnType<typeof loadAppProcessSuccess>
  | ReturnType<typeof loadAppProcessFailure>
  | ReturnType<typeof loadTopPage>
  | ReturnType<typeof loadTopPageSuccess>
  | ReturnType<typeof loadTopPageFailure>
  | ReturnType<typeof loadSagaPage>
  | ReturnType<typeof loadSagaPageSuccess>
  | ReturnType<typeof loadSagaPageFailure>
  | ReturnType<typeof loadApolloPage>
  | ReturnType<typeof loadApolloPageSuccess>
  | ReturnType<typeof loadApolloPageFailure>;

export const SET_ENV = 'SET_ENV';
export const setEnv = (env: State['env']) =>
  ({
    type: SET_ENV,
    payload: {
      env
    }
  } as const);

export const SET_BASE_URL = 'SET_BASE_URL';
export const setBaseUrl = (baseUrl: State['baseUrl']) =>
  ({
    type: SET_BASE_URL,
    payload: {
      baseUrl
    }
  } as const);

export const RESET_PAGES_STATUS = 'RESET_PAGES_STATUS';
export const resetPageStatus = () =>
  ({
    type: RESET_PAGES_STATUS
  } as const);

export const LOAD_APP_PROCESS = 'LOAD_APP_PROCESS';
export const loadAppProcess = () =>
  ({
    type: LOAD_APP_PROCESS
  } as const);

export const LOAD_APP_PROCESS_SUCCESS = 'LOAD_APP_PROCESS_SUCCESS';
export const loadAppProcessSuccess = () =>
  ({
    type: LOAD_APP_PROCESS_SUCCESS
  } as const);

export const LOAD_APP_PROCESS_FAILURE = 'LOAD_APP_PROCESS_FAILURE';
export const loadAppProcessFailure = (err: Error) =>
  ({
    type: LOAD_APP_PROCESS_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_TOP_PAGE = 'LOAD_TOP_PAGE';
export const loadTopPage = () =>
  ({
    type: LOAD_TOP_PAGE
  } as const);

export const LOAD_TOP_PAGE_SUCCESS = 'LOAD_TOP_PAGE_SUCCESS';
export const loadTopPageSuccess = () =>
  ({
    type: LOAD_TOP_PAGE_SUCCESS
  } as const);

export const LOAD_TOP_PAGE_FAILURE = 'LOAD_TOP_PAGE_FAILURE';
export const loadTopPageFailure = (err: Error) =>
  ({
    type: LOAD_TOP_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_SAGA_PAGE = 'LOAD_SAGA_PAGE';
export const loadSagaPage = (maxLength: string | null) =>
  ({
    type: LOAD_SAGA_PAGE,
    payload: {
      maxLength
    }
  } as const);

export const LOAD_SAGA_PAGE_SUCCESS = 'LOAD_SAGA_PAGE_SUCCESS';
export const loadSagaPageSuccess = () =>
  ({
    type: LOAD_SAGA_PAGE_SUCCESS
  } as const);

export const LOAD_SAGA_PAGE_FAILURE = 'LOAD_SAGA_PAGE_FAILURE';
export const loadSagaPageFailure = (err: Error) =>
  ({
    type: LOAD_SAGA_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);

export const LOAD_APOLLO_PAGE = 'LOAD_APOLLO_PAGE';
export const loadApolloPage = () =>
  ({
    type: LOAD_APOLLO_PAGE
  } as const);

export const LOAD_APOLLO_PAGE_SUCCESS = 'LOAD_APOLLO_PAGE_SUCCESS';
export const loadApolloPageSuccess = () =>
  ({
    type: LOAD_APOLLO_PAGE_SUCCESS
  } as const);

export const LOAD_APOLLO_PAGE_FAILURE = 'LOAD_APOLLO_PAGE_FAILURE';
export const loadApolloPageFailure = (err: Error) =>
  ({
    type: LOAD_APOLLO_PAGE_FAILURE,
    payload: {
      err
    }
  } as const);
