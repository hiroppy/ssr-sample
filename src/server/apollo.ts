import * as express from 'express';
import { ApolloServer } from 'apollo-server-express';
import { getComplexity, simpleEstimator, fieldExtensionsEstimator } from 'graphql-query-complexity';
import { separateOperations } from 'graphql';
import { resolvers, schema } from '../graphql/schema';
import { endpoint, limitCost } from '../graphql/constants';

export function apollo(app: express.Application) {
  const apollo = new ApolloServer({
    debug: process.env.NODE_ENV !== 'production',
    schema,
    resolvers,
    playground:
      process.env.NODE_ENV !== 'production'
        ? {
            endpoint
          }
        : false,
    plugins: [
      {
        requestDidStart: () => ({
          didResolveOperation({ request, document }) {
            const complexity = getComplexity({
              schema,
              query: request.operationName
                ? separateOperations(document)[request.operationName]
                : document,
              variables: request.variables,
              estimators: [fieldExtensionsEstimator(), simpleEstimator({ defaultComplexity: 1 })]
            });

            if (complexity >= limitCost) {
              throw new Error(`${complexity} is over ${limitCost}`);
            }
            console.log('Used query complexity points:', complexity);
          },
          didEncounterErrors({ errors, request }) {
            // e.g. add a logger
            console.error(errors);
            console.error(request);
          }
        })
      }
    ]
  });

  apollo.applyMiddleware({ app, path: endpoint });
}
