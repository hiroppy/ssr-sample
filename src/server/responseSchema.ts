import raw from 'raw.macro';

export type Samples = typeof apolloSamples;

// export type Themes = {
//   [key in 'dark' | 'light']: {
//     main: string;
//     sub: string;
//   };
// };

// TODO: fix
// need to create babel-plugin

export const sagaSamples = [
  {
    id: 1,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code: process.env.NODE_ENV !== 'test' ? raw('../client/sagas/pages.ts') : 'test-code',
    name: 'pages.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];

export const apolloSamples = [
  {
    id: 1,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 2,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 3,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  },
  {
    id: 4,
    code:
      process.env.NODE_ENV !== 'test'
        ? raw('../client/components/pages/Apollo/Apollo.tsx')
        : 'test-code',
    name: 'apollo.tsx',
    likeCount: 0,
    description: 'aaaa'
  }
];
