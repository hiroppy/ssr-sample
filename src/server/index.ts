import { join } from 'path';
import cluster from 'cluster';
import { cpus } from 'os';
import { config } from 'dotenv';
import { runServer } from './server';

const isProd = process.env.NODE_ENV === 'production';

// If you compile server code with webpack, this is unnecessary.
config({
  path: isProd ? join(__dirname, '../../../.env.prod') : join(__dirname, '../../.env.dev')
});

if (isProd) {
  const numCPUs = cpus().length;

  if (cluster.isMaster) {
    [...new Array(numCPUs)].forEach(() => cluster.fork());

    // cluster manager
    cluster.on('exit', (worker, code, signal) => {
      console.log(`Restarting ${worker.process.pid}. ${code || signal}`);
      cluster.fork();
    });
  } else {
    runServer();
  }
} else {
  runServer();
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

process.on('unhandledRejection', (err) => {
  console.error(err);
});
