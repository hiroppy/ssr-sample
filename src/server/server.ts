/* eslint @typescript-eslint/no-var-requires: 0 */

import { createServer } from 'http';
import express from 'express';
import * as bodyParser from 'body-parser';
import compression from 'compression';
import { generateNonceId, csp } from './csp';
import { apollo } from './apollo';
import { router } from './router';

export function runServer() {
  const app = express();
  const port = process.env.PORT || 3000;

  app.use(bodyParser.urlencoded({ extended: true }));
  app.use(bodyParser.json());

  // CSP
  app.use(generateNonceId);
  app.use(csp);

  // compression
  app.use(compression({ level: 5 }));

  // HMR
  if (process.env.NODE_ENV === 'development') {
    const webpack = require('webpack');
    const webpackHotMiddleware = require('webpack-hot-middleware');
    const webpackDevMiddleware = require('webpack-dev-middleware');
    const config = require('../../webpack.client.config');
    const compiler = webpack(config);

    app.use(webpackHotMiddleware(compiler));
    app.use(
      webpackDevMiddleware(compiler, {
        publicPath: config.output.publicPath,
        writeToDisk(filePath: string) {
          return /loadable-stats/.test(filePath);
        }
      })
    );
  }

  // setup apollo
  apollo(app);

  // register routes
  router(app);

  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'test') {
    const server = createServer(app);

    server.listen(port, () => {
      console.log(`Listening on ${port}`);
    });

    // https://stackoverflow.com/a/48710483/7014700
    server.on('error', (err: NodeJS.ErrnoException) => {
      if (err.syscall !== 'listen') throw err;

      const bind = typeof port === 'string' ? `Pipe ${port}` : `Port ${port}`;

      switch (err.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw err;
      }
    });
  }

  return app;
}
