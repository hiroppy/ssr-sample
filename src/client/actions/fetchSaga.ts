import { loadSagaPage } from './pages';
import { sagaSamples } from '../../server/responseSchema';

export const FETCH_SAGA_CODE = 'FETCH_SAGA_CODE';
export const fetchSagaCode = (maxLength: Parameters<typeof loadSagaPage>[0]) =>
  ({
    type: FETCH_SAGA_CODE,
    payload: {
      maxLength
    }
  } as const);

export const FETCH_SAGA_CODE_SUCCESS = 'FETCH_SAGA_CODE_SUCCESS';
export const fetchSagaCodeSuccess = (samples: typeof sagaSamples) =>
  ({
    type: FETCH_SAGA_CODE_SUCCESS,
    payload: {
      samples
    }
  } as const);

export const FETCH_SAGA_CODE_FAILURE = 'FETCH_SAGA_CODE_FAILURE';
export const fetchSagaCodeFailure = (err: Error) =>
  ({
    type: FETCH_SAGA_CODE_FAILURE,
    payload: {
      err
    }
  } as const);

export const ADD_LIKE = 'ADD_LIKE';
export const addLike = (id: number) =>
  ({
    type: ADD_LIKE,
    payload: {
      id
    }
  } as const);

export const ADD_LIKE_SUCCESS = 'ADD_LIKE_SUCCESS';
export const addLikeSuccess = (res: typeof sagaSamples[0]) =>
  ({
    type: ADD_LIKE_SUCCESS,
    payload: {
      res
    }
  } as const);

export const ADD_LIKE_FAILURE = 'ADD_LIKE_FAILURE';
export const addLikeFailure = (err: Error) =>
  ({
    type: ADD_LIKE_FAILURE,
    payload: {
      err
    }
  } as const);

export type Actions =
  | ReturnType<typeof fetchSagaCode>
  | ReturnType<typeof fetchSagaCodeSuccess>
  | ReturnType<typeof fetchSagaCodeFailure>
  | ReturnType<typeof addLike>
  | ReturnType<typeof addLikeSuccess>
  | ReturnType<typeof addLikeFailure>;
