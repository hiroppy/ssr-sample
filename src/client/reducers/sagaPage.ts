import { Actions } from '../actions/fetchSaga';
import { Samples } from '../../server/responseSchema';

export type State = {
  samples: Samples;
};

export const initialState: State = {
  samples: []
};

export const reducer = (state: State = initialState, action: Actions): State => {
  switch (action.type) {
    case 'FETCH_SAGA_CODE_SUCCESS':
      return { ...state, samples: action.payload.samples };
    case 'FETCH_SAGA_CODE':
    case 'FETCH_SAGA_CODE_FAILURE':
    default:
      return state;
  }
};
