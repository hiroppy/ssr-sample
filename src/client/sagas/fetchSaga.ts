import { put, call, takeLatest, select } from 'redux-saga/effects';
import fetch from 'cross-fetch';
import {
  FETCH_SAGA_CODE,
  fetchSagaCode as FetchSagaCode,
  fetchSagaCodeSuccess,
  fetchSagaCodeFailure,
  ADD_LIKE,
  addLike as AddLike,
  addLikeSuccess,
  addLikeFailure
} from '../actions/fetchSaga';
import { State } from '../reducers/sagaPage';
import { getBaseUrl } from '../selectors';

function* fetchMethod({
  path,
  params,
  options
}: {
  path: string;
  params?: Record<string, any>;
  options?: Record<string, any>;
}) {
  const baseUrl = yield select(getBaseUrl);
  const url = new URL(`${baseUrl || window.origin}${path}`);

  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      url.searchParams.append(key, value || '');
    });
  }

  const res: Response = yield call(fetch, url.toString(), options);
  const { body }: { body: State['samples'] } = yield res.json();

  return body;
}

function* fetchSagaCode(action: ReturnType<typeof FetchSagaCode>) {
  try {
    const params = { maxLength: action.payload.maxLength };
    const body = yield fetchMethod({
      path: '/api/saga',
      params
    });

    yield put(fetchSagaCodeSuccess(body));
  } catch (err) {
    yield put(fetchSagaCodeFailure(err));
  }
}

function* addLike(action: ReturnType<typeof AddLike>) {
  const { id } = action.payload;
  const options = { method: 'POST' };

  try {
    // actually, this method doesn't run on Node.js so it's unnecessary to create a full url
    const body = yield fetchMethod({
      path: `/api/saga/${id}`,
      options
    });

    // need to optimize this because it's better to avoid fetching all data
    // currently, apollo's mutation uses `refetchQueries` so make the behavior the same.
    yield put(FetchSagaCode(null));
    yield put(addLikeSuccess(body));
  } catch (err) {
    yield put(addLikeFailure(err));
  }
}

export function* fetchSagaProcess() {
  yield takeLatest(FETCH_SAGA_CODE, fetchSagaCode);
  yield takeLatest(ADD_LIKE, addLike);
}
