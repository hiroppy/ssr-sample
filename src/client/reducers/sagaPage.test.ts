import { reducer, initialState } from './sagaPage';
import * as actions from '../actions/fetchSaga';
import { sagaSamples } from '../../server/responseSchema';

test('should return the initial state', () => {
  expect(reducer(undefined, { type: undefined } as any)).toEqual(initialState);
});

test('should handle FETCH_SAGA_CODE', () => {
  expect(reducer(initialState, actions.fetchSagaCode('2'))).toMatchSnapshot();
});

test('should handle FETCH_SAGA_CODE_SUCCESS', () => {
  expect(reducer(initialState, actions.fetchSagaCodeSuccess(sagaSamples))).toMatchSnapshot();
});

test('should handle FETCH_SAGA_CODE_FAILURE', () => {
  expect(reducer(initialState, actions.fetchSagaCodeFailure(new Error('404')))).toMatchSnapshot();
});
