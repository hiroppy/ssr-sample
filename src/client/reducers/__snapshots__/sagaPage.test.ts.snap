// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle FETCH_SAGA_CODE 1`] = `
Object {
  "samples": Array [],
}
`;

exports[`should handle FETCH_SAGA_CODE_FAILURE 1`] = `
Object {
  "samples": Array [],
}
`;

exports[`should handle FETCH_SAGA_CODE_SUCCESS 1`] = `
Object {
  "samples": Array [
    Object {
      "code": "test-code",
      "description": "aaaa",
      "id": 1,
      "likeCount": 0,
      "name": "pages.tsx",
    },
    Object {
      "code": "test-code",
      "description": "aaaa",
      "id": 2,
      "likeCount": 0,
      "name": "pages.tsx",
    },
    Object {
      "code": "test-code",
      "description": "aaaa",
      "id": 3,
      "likeCount": 0,
      "name": "pages.tsx",
    },
    Object {
      "code": "test-code",
      "description": "aaaa",
      "id": 4,
      "likeCount": 0,
      "name": "pages.tsx",
    },
  ],
}
`;
